mne/beamformer/_dics.py

```diff
@@ -1,32 +1,482 @@
+
+    Notes
+    -----
+    The original reference is [1]_.
+
+    For more information about ``real_filter``, see the
+    `supplemental information <http://www.cell.com/cms/attachment/616681/4982593/mmc1.pdf>`_
+    from [3]_.
+
+    References
+    ----------
+    .. [1] Gross et al. Dynamic imaging of coherent sources: Studying neural
+           interactions in the human brain. PNAS (2001) vol. 98 (2) pp. 694-699
+    .. [2] Sekihara & Nagarajan. Adaptive spatial filters for electromagnetic
+           brain imaging (2008) Springer Science & Business Media
+    .. [3] Hipp JF, Engel AK, Siegel M (2011) Oscillatory Synchronization
+           in Large-Scale Cortical Networks Predicts Perception.
+           Neuron 69:387-396.
+    """  # noqa: E501
+    allowed_ori = [None, 'normal', 'max-power']
+    if pick_ori not in allowed_ori:
+        raise ValueError('"pick_ori" should be one of %s.' % allowed_ori)
+
+    frequencies = [np.mean(freq_bin) for freq_bin in csd.frequencies]
+    n_freqs = len(frequencies)
+    n_orient = forward['sol']['ncol'] // forward['nsource']
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori,
+        leadfield_norm=leadfield_norm,
+    )
+    csd_picks = [csd.names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_matrix(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # Normalize the spatial filters
+            if Wk.ndim == 2 and len(Wk) > 1:
+                # Free source orientation
+                # Officially, the formula calls for inv(Ck) @ Wk.
+                # However, since Ck is most likely rank deficient, the
+                # following is more stable in practice.
+                Wk /= np.diag(Ck)[:, np.newaxis]
+            else:
+                # Fixed source orientation
+                Wk /= Ck
+
+            if pick_ori == 'max-power':
```

Comment:
Note that there is a different formula for orientation selection when performing weight normalization (cf. ``lcmv`` code).

Reply:
I'm actually not sure that formula 4.47 of Sekihara & Nagarajan doesn't holds up when you apply regularization to the covariance/csd matrix. That formula assumes that `Cm == inv(Cm_inv)`, however, this is not the case if we compute `Cm_inv` as `inv(Cm + reg * np.eye(n))`. This may be why we get different results when using formula 4.47 as opposed to scaling the weights at the end of the function as we do for the other cases.

Reply:
Judging from the optimization criterion 4.11 in that book and the way they derive the unit-gain beamfomer, our approach of first computing the weights and then scaling them to assure `W @ W.T == np.eye(n)` is valid and should yield the same results as what formula 4.47 is trying to compute.

Reply:
They explicitly mention `tau` as scaling factor. It's just a matter of computing it.

Reply:
I"ve make a version that incorporates the formula used in LCMV (copy-pasting the code from `_lcmv.py` for this). The results look fine on the simulation:
![fig1](https://user-images.githubusercontent.com/428273/34207542-979ca582-e593-11e7-9c83-f17e115700ad.png)

However, the resulting filter is *not* unit-noise-gain, as
`filters['weights'][0][0] @ filters['weights'][0][0].T != 1` (it's ~0.66, not even close)

Reply:
@wmvanvliet It's possible that the noise gain needs to be scaled to actually become unity, but the important point is the noise gain is equal across voxels. Is that the case?

Reply:
Oops, in the figure above, I had `recude_rank=True` set and this was unnecessary. With `reduce_rank=False` the picture becomes very similar to the version that scales the filter weighs at the end of the function:

![fig2](https://user-images.githubusercontent.com/428273/34487872-d93c67d4-efde-11e7-89af-8033d06ef31d.png)

Reply:
@sarangnemo However, the weights are still not unit-noise-gain. Here is `W @ W.T` plotted along the cortex.  It is pretty smooth, but not exactly:

![fig1](https://user-images.githubusercontent.com/428273/34487949-227e2216-efdf-11e7-9a0e-72676aa3eada.png)

Reply:
If anyone wants to try the two different approaches side-by-side, I invite them to checkout the `new_formula` branch of my MNE-Python fork: https://github.com/wmvanvliet/mne-python/tree/new_formula

Reply:
@wmvanvliet it seems that there's been an enormous point of confusion here. You've actually implemented a vector beamformer. :-) This is indeed the textbook way of doing that (including the strategy of taking the diagonal in the normalization step -- see section 4.5.3 and equation 4.67 of Sekihara & Nagarajan). Finding the optimal orientation for a vector source is not dependent on the particular beamformer formula, and what you have done should work.

However, an optimal scalar beamformer approach should provide results that are nearly identical, and even if a non-optimal orientation ends up being selected, this should not violate unit-noise-gain constraint. You should be able to take the optimal orientation as computed in @britta-wstnr 's scalar beamformer code, modify your lead field with it, and proceed with the rest of your code. At the very least, the noise gain should indeed be 1, and with any luck, the results will be (almost) the same as what you get with a vector beamformer and selecting orientation post-hoc.

Reply:
Besides, i checked the unit-noise-gain BF in ``_lcmv.py`` with one of the example data sets, and the noise gain is 1 there - i.e., with time domain data.

Reply:
@britta-wstnr You're right about that. I must have made some sort of mistake when I transferred to code to `_dics.py`.